src/System.Net.Security/src/System/Net/Security/SslApplicationProtocol.cs

```diff
@@ -0,0 +1,146 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Text;
+
+namespace System.Net.Security
+{
+    public struct SslApplicationProtocol : IEquatable<SslApplicationProtocol>
+    {
+        private readonly byte[] _protocol;
+        private readonly ReadOnlyMemory<byte> _readOnlyProtocol;
+        private static readonly Encoding s_utf8 = Encoding.GetEncoding(Encoding.UTF8.CodePage, EncoderFallback.ExceptionFallback, DecoderFallback.ExceptionFallback);
+
+        // h2
+        public static readonly SslApplicationProtocol Http2 = new SslApplicationProtocol(new byte[] { 0x68, 0x32 }, false);
+        // http/1.1
+        public static readonly SslApplicationProtocol Http11 = new SslApplicationProtocol(new byte[] { 0x68, 0x74, 0x74, 0x70, 0x2f, 0x31, 0x2e, 0x31 }, false);
+
+        internal SslApplicationProtocol(byte[] protocol, bool copy)
+        {
+            if (protocol == null)
+            {
+                throw new ArgumentNullException(nameof(protocol));
+            }
+
+            if (protocol.Length == 0)
```

Comment:
The WIndows and OpenSSL implementations both fail if a protocol of length > 255 bytes is used.  If that's a protocol limit, maybe it should also be captured here?

Reply:
Yes, that appears to be a protocol limit:
https://tools.ietf.org/rfc/rfc7301.txt
   opaque ProtocolName<1..2^8-1>;
Adding the same limit here would be good.